platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/JavaRecordCodec.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache.serialization.codecs
+
+import org.gradle.configurationcache.problems.PropertyKind
+import org.gradle.configurationcache.serialization.ReadContext
+import org.gradle.configurationcache.serialization.WriteContext
+import org.gradle.configurationcache.serialization.beans.readPropertyValue
+import org.gradle.configurationcache.serialization.beans.reportUnsupportedFieldType
+import org.gradle.configurationcache.serialization.beans.unsupportedFieldTypeFor
+import org.gradle.configurationcache.serialization.beans.writeNextProperty
+import org.gradle.configurationcache.serialization.withDebugFrame
+import java.lang.reflect.Field
+import java.lang.reflect.Modifier.isStatic
+
+
+object JavaRecordCodec : EncodingProducer, Decoding {
+
+    override fun encodingForType(type: Class<*>): Encoding? =
+        // need to check by name because it's only present in Java 14+
+        JavaRecordEncoding.takeIf { type.superclass?.canonicalName == "java.lang.Record" }
+
+
+    override suspend fun ReadContext.decode(): Any? {
+        val clazz = readClass()
+        val args = mutableListOf<Any?>()
+        for (field in clazz.relevantFields) {
+            val fieldName = field.name
+            unsupportedFieldTypeFor(field)?.let {
+                reportUnsupportedFieldType(it, "deserialize", fieldName)
+            }
+            readPropertyValue(PropertyKind.Field, fieldName) { fieldValue ->
+                args.add(fieldValue)
+            }
+        }
+        val constructor = clazz.constructors.find { it.parameterCount == args.size }
```

Comment:
❓ Should we also check for the types of the arguments? It is possible for record to have multiple constructors of the same arity.

Reply:
An official blog post from the JDK folks suggest to always deserialize via the primary constructor. However, in the snippet they provide, they actually find this constructor by the signature as @mlopatkin suggests.

https://inside.java/2021/04/06/record-serialization-in-practise/#:~:text=the%20deserialization%20of%20a%20record%20uses%20only%20the%20canonical%20constructor.

The snippet in the article takes a slightly different approach to (de)serialization using the new static methods. Could we also do that if we had a guard checking for the Java 14+ runtime?

Reply:
@mlopatkin Good point, thanks! I thought it would be ok since the primary constructor would be the first in the list of constructors, but found no hard proof in JVM spec for that, so better be safe than sorry. I've updated the PR.

@alllex, I think the current solution works fine, and the snippet does the similar stuff I do in the updated PR. I think relying on JDK14 overcomplicate things.